### DISCUSSION TITLE
g-svg umd script run error

### DISCUSSION THREAD

**Discussion by deyihu** [antvis/G2#4951] [Q&A] [answered]
##### Checklist

- [ ] `npm test` passes
- [ ] benchmarks are included
- [ ] commit message follows commit guidelines
- [ ] documents are updated

##### Description of change

- g-svg umd script run error 

```html
<script src="https://unpkg.com/@antv/g-svg@1.8.55/dist/index.umd.min.js"></script>
```
![2023-05-05_141716](https://user-images.githubusercontent.com/25998927/236388859-2ed566c1-050d-4b42-9493-324a1cd6bdf5.png)

- Issue Helper page error
 [Issue Helper](https://antv-issue-helper.surge.sh/)

![2023-05-05_142047](https://user-images.githubusercontent.com/25998927/236389251-7c12f65e-8926-4c60-b5b1-7483241da7a2.png)

- the demo about  SVGRenderer Not the expected result
 [svg render](https://g2.antv.antgroup.com/examples/renderer/renderer/#svg) the render result still canvas 

![2023-05-05_142546](https://user-images.githubusercontent.com/25998927/236390013-18cf06fc-7ac7-4db6-a835-6f42c28a2b79.png)

**Comment by xiaoiver** [comment, marked as answer]
UMD 格式 `g-svg` 的使用方式如下：

```html
<script src="https://unpkg.com/@antv/g"></script>
<script src="https://unpkg.com/@antv/g-svg"></script>
<script src="https://unpkg.com/@antv/g2@5.0.5/dist/g2-lite.min.js"></script>
```

命名空间为 `G.SVG`：

```js
const svgRenderer = new window.G.SVG.Renderer();

const { Chart } = window.G2;
const chart = new Chart({
  container: 'container',
  theme: 'classic',
  renderer: svgRenderer // 传入 SVG Renderer
});
```

在线示例：https://codepen.io/xiaoiver/pen/zYmpbNJ

### 什么情况下使用 g2-lite 版本？

正常情况下使用 g2 UMD 即可，里面已经把 `@antv/g` 和 `@antv/g-canvas` 打包进去了。
但是当我们想使用非 Canvas 渲染器、g 提供的其他插件，就需要使用 g2-lite UMD，理由如下。

g-svg 这样的渲染器以及其他 g 的插件需要从 window.G 这样的命名空间下获取信息，而 g2 UMD 并不会暴露，这也是上面截图报错 undefined 的原因。

而 g2-lite UMD 版本会 external 掉 `@antv/g` 和 `@antv/g-canvas`，因此在使用时需要引入它俩的 UMD。

**Reply by deyihu** [reply]
@xiaoiver 
lite版本我试了，部分图表还是有问题的，
![2023-05-07_121036](https://user-images.githubusercontent.com/25998927/236657203-df3e1089-8da6-495e-8a62-207fbe47cc18.png)

不知道是不是G2内部写死导致，还是会去create Canvas renderer
https://codepen.io/deyihu/pen/zYmRqbp?editors=1010

理论上外部传入了， G2内部应该完全采用外部传入的renderer才对

```js
const chart = new Chart({
  container: 'container',
  theme: 'classic',
  renderer: svgRenderer // 传入 SVG Renderer
});
```

尴尬的是如果不使用lite版本，外部传入renderer完全不生效

**Reply by deyihu** [reply]
鉴于打包体积的考虑，提供lite版本可以理解，而我的理解是：

- 只要用户外部传入renderer就应该使用用户传入的renderer
- 不管用户用lite版本还是完全版本的都应该遵循该原则
- 只有当用户不传renderer时，G2内部才需要自己去create renderer
- 基于以上，目前G2的总体设计有点没看懂
